### PR TITLE
Protectionfilters

### PR DESCRIPTION
**Pull request by jbrichau**
Added `WASessionCookieProtectionFilter` to secure Seaside sessions from session hijacking while allowing to use multiple sessions in a single browser. 

An example was added: `WASessionProtectedCounterExample` to illustrate usage.

Makes a breaking API change by renaming `WAProtectionFilter` to `WARemoteAddressProtectionFilter`. It felt better this way. If we want to merge without this, we can undo the rename. In the other case, it can only be merged with a major version update.

ps: I want to push the merge button myself ;) I might want to add stuff before. But I'm already asking a review for this ;)

### CODE REVIEW

**Comment by marschall** [comment]
We should probably add a change not somewhere that `WAProtectionFilter` got renamed to `WARemoteAddressProtectionFilter`.

File: repository/Seaside-Core.package/WASessionCookieProtectionFilter.class/README.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+The session cookie protection filter ensures that the wrapped request handler only accepts requests from the same browser session. This filter is specifically useful to protect session hijacking when using the (default) query field session tracking strategy.
+
+Because WAQueryFieldHandlerTrackingStrategy puts the Seaside session key in the url, a session can be easily hijacked by copying the url. This request filter prevents this by requiring a browser session cookie associated to the Seaside session. As a result, a copied Seaside url can only be used in the same browser session.
```
Comment (marschall) [inline repository/Seaside-Core.package/WASessionCookieProtectionFilter.class/README.md]: I find the word "browser session" confusing. Is that an established term for all tabs and windows of a browser. Would "browser instance" be better or equally confusing.

File: repository/Seaside-Core.package/WASessionCookieProtectionFilter.class/class/browserSessionIDCookieKey.st [hunk 2]
```diff
@@ -0,0 +1,4 @@
+private - constants
+browserSessionIDCookieKey
+
+	^ 'seaside_browser_session'
```
Comment (marschall) [inline repository/Seaside-Core.package/WASessionCookieProtectionFilter.class/class/browserSessionIDCookieKey.st]: Why is this on the class side?

Comment (marschall) [inline repository/Seaside-Core.package/WASessionCookieProtectionFilter.class/class/browserSessionIDCookieKey.st]: Traditionally uses uses short and cryptic key names like `_b` but this is fine as well.

File: repository/Seaside-Core.package/WASessionCookieProtectionFilter.class/instance/privateBrowserSessionID..st [hunk 3]
```diff
@@ -0,0 +1,3 @@
+private
+privateBrowserSessionID: aValue
```
Comment (marschall) [inline repository/Seaside-Core.package/WASessionCookieProtectionFilter.class/instance/privateBrowserSessionID..st]: Traditionally we have not used the `private` prefix for private methods, we simply added them to the private protocol.

File: repository/Seaside-Examples.package/WASessionProtectedCounter.class/properties.json [hunk 4]
```diff
@@ -0,0 +1,11 @@
+{
+	"commentStamp" : "",
+	"super" : "WACounter",
```
Comment (marschall) [inline repository/Seaside-Examples.package/WASessionProtectedCounter.class/properties.json]: I don't think we should encourage subclassing for this. I would rather wrap a `WACounter` or add a decoration.

File: repository/Seaside-Core.package/WASessionCookieProtectionFilter.class/README.md (hunk 1 above)
Comment (jbrichau) [reply repository/Seaside-Core.package/WASessionCookieProtectionFilter.class/README.md]: Here's where it's coming from: we use a 'browser session cookie'. So, that's why we refer to it as a 'browser session'

File: repository/Seaside-Core.package/WASessionCookieProtectionFilter.class/class/browserSessionIDCookieKey.st (hunk 2 above)
Comment (jbrichau) [reply repository/Seaside-Core.package/WASessionCookieProtectionFilter.class/class/browserSessionIDCookieKey.st]: I move it to the instance side. It was used to have this key available from test methods in our codebase, but it makes no sense now.

Comment (jbrichau) [reply repository/Seaside-Core.package/WASessionCookieProtectionFilter.class/class/browserSessionIDCookieKey.st]: In the case of cookie names, we might want to be a bit less cryptic ;)

File: repository/Seaside-Core.package/WASessionCookieProtectionFilter.class/instance/privateBrowserSessionID..st (hunk 3 above)
Comment (jbrichau) [reply repository/Seaside-Core.package/WASessionCookieProtectionFilter.class/instance/privateBrowserSessionID..st]: Agreed; better to keep the code base consistent. I rename them.

File: repository/Seaside-Examples.package/WASessionProtectedCounter.class/properties.json (hunk 4 above)
Comment (jbrichau) [reply repository/Seaside-Examples.package/WASessionProtectedCounter.class/properties.json]: Ok. Will change

File: repository/Seaside-Core.package/WAQueryFieldHandlerTrackingStrategy.class/README.md [hunk 5]
```diff
@@ -4,4 +4,4 @@ I track request handlers using a query field. This results in URLs looking like
...
\ No newline at end of file
+However there are some drawbacks. The smallest is aesthetical, the URL in the address bar of the browser is less "clean". Second because the session id is part of the request URL it shows up in all kinds of places. For example server logs. Not only the log of the server running the application but also very web site visited from there because it shows up in the Referer HTTP header. The danger of this is that when somebody knows the session id of somebody else he can take over his session. Most of these problems can be mitigated by adding a WARemoteAddressProtectionFilter or WASessionCookieProtectionFilter to every session. The first one this creates new problems for users with changing IPs (eg. mobile devices).
```
Comment (theseion) [inline repository/Seaside-Core.package/WAQueryFieldHandlerTrackingStrategy.class/README.md]: The last sentence doesn't quite make sense.

Comment (jbrichau) [reply repository/Seaside-Core.package/WAQueryFieldHandlerTrackingStrategy.class/README.md]: oeps. thanks for catching that.

**Comment by theseion** [comment]
@jbrichau Did you see my comment? https://github.com/SeasideSt/Seaside/pull/1148#pullrequestreview-279335608

**Comment by jbrichau** [comment]
@theseion Ha, no, I didn't.
I think this is something for another issue, as I did not touch that codebase (except rename the class). 
Ok if I copy/paste this to another issue?

**Comment by theseion** [comment]
Yes, please.

**Pull request merged**